astropy/coordinates/baseframe.py

```diff
@@ -232,19 +300,49 @@ def __init__(self, *args, **kwargs):
             if repr_kwargs:
                 if repr_kwargs.get('distance', True) is None:
                     del repr_kwargs['distance']
+
                 if (issubclass(self.representation, SphericalRepresentation) and
                         'distance' not in repr_kwargs):
                     representation = self.representation._unit_representation
                 else:
                     representation = self.representation
                 representation_data = representation(copy=copy, **repr_kwargs)
 
+            # Now we handle the Differential data:
+            # Get any differential data passed in to the frame initializer
+            # using keyword or positional arguments for the component names
+            diff_kwargs = {}
+            for nmkw, nmrep in self.differential_component_names.items():
+                if len(args) > 0:
+                    # first gather up positional args
+                    diff_kwargs[nmrep] = args.pop(0)
+                elif nmkw in kwargs:
+                    diff_kwargs[nmrep] = kwargs.pop(nmkw)
+
+            if diff_kwargs:
+                if ((issubclass(self.differential_cls, BaseSphericalDifferential) or
+                     issubclass(self.differential_cls, BaseSphericalCosLatDifferential)) and
+                        hasattr(self.differential_cls, '_unit_differential') and
+                        'd_distance' not in diff_kwargs):
+                    differential = self.differential_cls._unit_differential
+
+                elif ((issubclass(self.differential_cls, BaseSphericalDifferential) or
+                       issubclass(self.differential_cls, BaseSphericalCosLatDifferential)) and
+                        len(diff_kwargs) == 1 and 'd_distance' in diff_kwargs):
+                    differential = RadialDifferential
+
+                else:
+                    differential = self.differential_cls
+                differential_data = differential(copy=copy, **diff_kwargs)
+
         if len(args) > 0:
             raise TypeError(
                 '{0}.__init__ had {1} remaining unhandled arguments'.format(
                     self.__class__.__name__, len(args)))
 
         self._data = representation_data
+        if self._data and differential_data is not None:
```

Comment:
Note to @eteq and self: this is why the differential data gets dropped if you do:

```python
ICRS(pm_ra=xx, pm_dec=xx, differential_cls=UnitSphericalDifferential)
```